openvdb/points/PointMove.h

```diff
@@ -0,0 +1,952 @@
+        void operator()(const AttributeArray& array) const
+        {
+            auto* handleAsInt = reinterpret_cast<AttributeHandle<int>*>(
+                new AttributeHandle<ValueT>(array));
+            mHandles.emplace_back(handleAsInt);
+        }
+
+    private:
+        HandleArray& mHandles;
+    }; // struct CacheHandleOp
+
+    template <typename LeafRangeT>
+    void cache(const LeafRangeT& range, const Index attributeIndex)
+    {
+        using namespace openvdb::math;
+
+        mHandles.clear();
+        CacheHandleOp op(mHandles);
+
+        for (auto leaf = range.begin(); leaf; ++leaf) {
+            auto& array = leaf->attributeArray(attributeIndex);
+            processTypedArray(array, op);
+        }
+    }
+
+private:
+    HandleArray mHandles;
+}; // struct AttributeHandles
+
+
+template <typename DeformerT, typename TreeT, typename FilterT>
+struct BuildMoveMapsOp
+{
+    using LeafT = typename TreeT::LeafNodeType;
+    using LeafArrayT = std::vector<LeafT*>;
+    using LeafManagerT = typename tree::LeafManager<TreeT>;
+
+    BuildMoveMapsOp(const DeformerT& deformer,
+                    GlobalPointIndexMap& globalMoveLeafMap,
+                    LocalPointIndexMap& localMoveLeafMap,
+                    const std::map<Coord, Index>& targetLeafMap,
+                    const math::Transform& targetTransform,
+                    const math::Transform& sourceTransform,
+                    const FilterT& filter)
+        : mDeformer(deformer)
+        , mGlobalMoveLeafMap(globalMoveLeafMap)
+        , mLocalMoveLeafMap(localMoveLeafMap)
+        , mTargetLeafMap(targetLeafMap)
+        , mTargetTransform(targetTransform)
+        , mSourceTransform(sourceTransform)
+        , mFilter(filter) { }
+
+    void operator()(LeafT& leaf, size_t idx) const
+    {
+        DeformerT deformer(mDeformer);
+        deformer.reset(leaf, idx);
+
+        // determine source leaf node origin and offset in the source leaf vector
+
+        Coord sourceLeafOrigin = leaf.origin();
+
+        auto sourceHandle = AttributeWriteHandle<Vec3f>::create(leaf.attributeArray("P"));
+
+        for (auto iter = leaf.beginIndexOn(mFilter); iter; iter++) {
+
+            const bool useIndexSpace = DeformerTraits<DeformerT>::IndexSpace;
+
+            // extract index-space position and apply index-space deformation (if applicable)
+
+            Vec3d positionIS = sourceHandle->get(*iter) + iter.getCoord().asVec3d();
+            if (useIndexSpace) {
+               deformer.apply(positionIS, iter);
+            }
+
+            // transform to world-space position and apply world-space deformation (if applicable)
+
+            Vec3d positionWS = mSourceTransform.indexToWorld(positionIS);
+            if (!useIndexSpace) {
+                deformer.apply(positionWS, iter);
+            }
+
+            // transform to index-space position of target grid
+
+            positionIS = mTargetTransform.worldToIndex(positionWS);
```

Comment:
cant you put the above lines into the !indexSpace scope? i.e. 
```
if(!useIndexSpace) {
    Vec3d positionWS = mSourceTerm.indexToWorld(positionIS);
    deformer.apply(positionWS, iter);
    positionIS = mTargetTransform.worldToIndex(positionWS);
}
```

Reply:
@richhones - you still need to apply the world->index->world transformation if the source and target transforms don't match, so as to put these positions into the index-space of the target. Note that there are other optimisations that could be introduced here, ignoring this world->index->world transformation if the transforms are the same as well as collapsing these two transformations down into a single transform or matrix if they're not. Neither of which I introduced because the world-space deformer is by far the most common, the index-space deformation is currently just an option and is only used in the unit tests. If that were to change, this can be optimised later without affecting ABI or the public API.